### PR TITLE
8317697: refactor-encapsulate x86 VM_Version::CpuidInfo

### PR DESCRIPTION
**Pull request by jankratochvil**
In OpenJDK project CRaC I had a [need to fetch new CpuidInfo without affecting the existing one](https://github.com/openjdk/crac/commit/ed4ad9ba31b77732dcede2eb743b2f389ec9a0fe#diff-6ed856c57ddbe33e49883adb7c52ec51ed377e5f697dfd6d8bea505a97bfc5a5R2743).
Which led me to encapsulate the object more and I think this no-functionality-change patch is even appropriate for JDK.
I am sure interested primarily to reduce the CRaC patchset boilerplate.

---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8317697](https://bugs.openjdk.org/browse/JDK-8317697): refactor-encapsulate x86 VM_Version::CpuidInfo (**Enhancement** - P4)

### Reviewers
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**) ⚠️ Review applies to [beee275a](https://git.openjdk.org/jdk/pull/16093/files/beee275afbd8a107b945cb8ea114ae02ac233bdd)
 * [Vladimir Kozlov](https://openjdk.org/census#kvn) (@vnkozlov - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk.git pull/16093/head:pull/16093` \
`$ git checkout pull/16093`

Update a local copy of the PR: \
`$ git checkout pull/16093` \
`$ git pull https://git.openjdk.org/jdk.git pull/16093/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 16093`

View PR using the GUI difftool: \
`$ git pr show -t 16093`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/16093.diff">https://git.openjdk.org/jdk/pull/16093.diff</a>

</details>

### Webrev
[Link to Webrev Comment](https://git.openjdk.org/jdk/pull/16093#issuecomment-1751781998)

### CODE REVIEW

**Review by dholmes-ora — COMMENTED** [review]
I don't hate this :) but I don't like seeing methods on structs. Should we make `CpuidInfo` a full-fledged class instead?

File: src/hotspot/cpu/x86/vm_version_x86.hpp [hunk 1]
```diff
@@ -524,36 +524,49 @@ class VM_Version : public Abstract_VM_Version {
     int          zmm_save[16*4]; // Save zmm0, zmm7, zmm8, zmm31
   };
 
+  class CpuidInfo : public _CpuidInfo {
```
Comment (dholmes-ora) [inline src/hotspot/cpu/x86/vm_version_x86.hpp:527]: Why not just declare the original `CpuidInfo` as a class instead of extending the struct ???

Comment (jankratochvil) [reply src/hotspot/cpu/x86/vm_version_x86.hpp:527]: This way the members are always zero-initialized. Which simplifies existing code
```
-VM_Version::CpuidInfo VM_Version::_cpuid_info = { 0, };
+VM_Version::CpuidInfo VM_Version::_cpuid_info;
```
And makes it more foolproof - therefore also fixing an existing bug (I haven't been aware of so far) in my code in the [CRaC](https://openjdk.org/projects/crac/) branch.

Although the ` = { 0, }` initialization above was not needed as it is a static member anyway.

That means I should change the name of this patch as it is no longer just a "refactor" ization.

Or should I stop the zero-initialization?

Comment (dholmes-ora) [reply src/hotspot/cpu/x86/vm_version_x86.hpp:527]: Sorry I'm not familiar enough with C++ initialization rules for structs versus classes here. IIUC:
- the original  struct instance was explicitly zero-initialized. 
- If we just made it a class there would not be any zero initialization
- If the class extends the struct then we (somehow) regain zero initialization
??

Maybe methods on a struct weren't so bad after all. I'd like to hear other views on this code change. (Just to be clear 2 reviews are needed for hotspot code changes anyway.)

Thanks

Comment (jankratochvil) [reply src/hotspot/cpu/x86/vm_version_x86.hpp:527]: >     * the original  struct instance was explicitly zero-initialized.

It had to be zero-initialized by user of the struct. Which is error prone if you forget the zero-initialization during each use of the struct type.

>     * If we just made it a class there would not be any zero initialization

struct vs. class does not matter.

>     * If the class extends the struct then we (somehow) regain zero initialization
>       ??

zero-initialized:
```
echo 'struct A { long a; }; void f() { A a=A(); }'|g++ -c -o 1.o -Wall -x c++ -;objdump -dC 1.o
```
NOT zero-initialized:
```
echo 'struct A { long a; }; void f() { A a    ; }'|g++ -c -o 1.o -Wall -x c++ -;objdump -dC 1.o
```
It is important if one calls the default constructor or not. With one inheritance one can enforce the zero-initialization by explicitly calling the inner `_CpuidInfo` default-constructor from the outer class `CpuidInfo`.

> Maybe methods on a struct weren't so bad after all. I'd like to hear other views on this code change. (Just to be clear 2 reviews are needed for hotspot code changes anyway.)

struct vs. class does not matter here. I am just mixing two refactorizations together. I can remove the zero-initialization (I can submit it as a separate merge request but then I do not need that part myself).

**Review by dholmes-ora — APPROVED** [review]
Okay this basic refactoring seems simpler/clearer to me. Thanks.

**Review by vnkozlov — APPROVED** [review]
Looks hood to me too.

**Comment by vnkozlov** [comment]
I submitted our internal testing before sponsoring.

**Pull request closed without merging**